[luxon] Minor fixes and improvements

hugofpsilva opened the pull request:
### Luxon
#### DateTime
Added overloads for local() and utc(). (Are now necessary due to the object param after the number params).
Marked several params as optional where Luxon doesn't throw errors.
Fixed typing errors in ordinal getter, resolvedLocaleOptions, toLocaleString, toLocaleParts, and all format presets.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://moment.github.io/luxon/api-docs/index.html

peterblazejewicz commented:
one topic at time, why those should be optional? Is that how the Luxon states this (in Docs)?
There is no single test-case e.g. for `fromJSDate` that covers no input.
I'd be cautious with such hanges, I've to fix a runtime by introduced by minor change in `markdown-it` recently, when `undefined` was just added to DT as proper input a month ago. There was no single test in native JS package to cover that scenario, so code started failing at runtime when people used 'undefined' as a valid value.
I'd say Luxon docs are clear to say 'required' for those params (assuming we are using JSDocs convention)

hugofpsilva commented:
You are again correct, the docs don't specify these arguments as optional. Then again, the docs aren't nearly 100% accurate. Just as a fast example, the local() documentation doesn't even mention the options object as a last parameter (but does show it in the examples).

IMO Luxon is very permissive with the values passed to some (not all) of its methods.

In the example you gave of `fromJSDate`, Luxon checks if the `date` param is a JS Date with 
`Object.prototype.toString.call(date) === "[object Date]";`
which, although far from perfect, by the spec available [here](https://tc39.es/ecma262/multipage/fundamental-objects.html#sec-object.prototype.tostring) should return `false` for `undefined` as `Object.prototype.toString.call(undefined)` returns `"[object Undefined]"`.

I also added type tests for all the methods for which I marked the arguments as optional.

Don't get me wrong, I agree with you that `DateTime.fromJSDate()` with no arguments makes no sense, and the others I changed make no sense as well, but it is indeed accepted by the source code and it does return a `DateTime`, although an invalid one, instead of throwing an error. If the source allows it, why shouldn't we?

CarsonF commented:
I agree with @peterblazejewicz here. Just because a JS library doesn't break with a certain input doesn't mean it's expected/"allowed". As you said, calling `from*` without an input doesn't make sense. Therefore TS can help enforce this, even beyond what the author has expressed / coded in runtime checks.

hugofpsilva commented:
Alright it actually makes a lot more sense to me as it was, so I'm down for keeping it, I just thought it wasn't TS's job to restrict the user beyond typechecking.

I'll roll back that part 😃

peterblazejewicz commented on types/luxon/src/datetime.d.ts:
```diff
@@ -961,7 +1027,7 @@ export class DateTime {
      *                                 //=> ]
      */
-    toLocaleParts(opts: DateTimeFormatOptions): Intl.DateTimeFormat[];
+    toLocaleParts(opts?: DateTimeFormatPreset | DateTimeFormatOptions): ToLocalePartsOutput;
```
So here there was just a wrong return, why not amend to `Intl.DateTimeFormatPart[]`?

hugofpsilva replied:
Absolutely, I didn't thoroughly check Intl for an equivalent. Amending.

peterblazejewicz commented on types/luxon/src/datetime.d.ts:
```diff
@@ -642,7 +708,7 @@ export class DateTime {
      * DateTime.local(2017, 5, 25).ordinal //=> 145
      */
-    get ordinal(): number | DateTime;
+    get ordinal(): number;
```
this one was ok, the change is breaking I believe, `number` for NaN, otherwise date time like object:
https://github.com/moment/luxon/blob/0cbffd9c87eca6ac9a75dce124ce890dd945bc78/src/impl/conversions.js#L88

hugofpsilva replied:
The get ordinal() always returns the ordinal property of that object, which is always a number
https://github.com/moment/luxon/blob/d95ddf12b3d9b116054a8c715c4aa596c4a34454/src/datetime.js#L1084
https://github.com/moment/luxon/blob/d95ddf12b3d9b116054a8c715c4aa596c4a34454/src/impl/conversions.js#L28

peterblazejewicz replied:
yes, my bad, sorry!

peterblazejewicz reviewed (approved):
LGTM!

hugofpsilva commented:
Ready to merge

mister-what commented on types/luxon/src/datetime.d.ts:
```diff
@@ -146,6 +147,19 @@ export interface DateObjectUnits {
 
 export type ConversionAccuracy = 'casual' | 'longterm';
 
+export type DateTimeFormatPresetValue = 'numeric' | 'short' | 'long';
+export interface DateTimeFormatPreset {
+    year?: DateTimeFormatPresetValue;
+    month?: DateTimeFormatPresetValue;
+    day?: DateTimeFormatPresetValue;
+    weekday?: DateTimeFormatPresetValue;
+    hour?: DateTimeFormatPresetValue;
+    minute?: DateTimeFormatPresetValue;
+    second?: DateTimeFormatPresetValue;
+    timeZoneName?: DateTimeFormatPresetValue;
+    hourCycle?: 'h23';
+}
```
This type is inherently broken.
The type allows:
```ts
DateTime.fromISO("2021-09-13T07:52:27.697Z").toLocaleString({
  ...DateTime.DATETIME_MED,
  hour: "short"
})
```
But crashes at runtime. `DateTimeFormatPresetValue` does not apply for all options in `DateTimeFormatPreset`.

joshuajaco replied:
It is not restricted enough in some cases and too restrictive in other.
For example the following throws a type error but works correctly at runtime:
```js
DateTime.fromISO("2021-09-13T07:52:27.697Z").toLocaleString({
  ...DateTime.DATETIME_MED,
  hour: "2-digit"
})
```

peterblazejewicz replied:
to be eventually fixed here: #55750

joshuajaco commented on types/luxon/src/datetime.d.ts:
```diff
@@ -943,7 +1001,7 @@ export class DateTime {
      * DateTime.now().toLocaleString({ hour: '2-digit', minute: '2-digit', hourCycle: 'h23' }); //=> '11:32'
      */
-    toLocaleString(formatOpts?: DateTimeFormatOptions, opts?: LocaleOptions): string;
+    toLocaleString(formatOpts?: DateTimeFormatPreset | DateTimeFormatOptions, opts?: LocaleOptions): string;
```
Why was this type changed at all? Should the presets always be valid `DateTimeFormatOptions` as well?

peterblazejewicz replied:
I believe the intent was to provide better typing for presets. I've read a book, the author rewrote ending 39 times, this applies to our code IMO
#55750
